coding_systems/dmd/coding_system.py

```diff
@@ -11,6 +11,34 @@ class CodingSystem(BaseCodingSystem):
         "term": ["term", "dmd_name", "name", "nm", "description"],
     }
 
+    def ancestor_relationships(self, codes):
+        amps = AMP.objects.using(self.database_alias).filter(id__in=codes)
+
+        # get VMPs that are either in `codes` or are ancestors
+        # of the AMPs in `codes`
+        codes = set(codes) | {amp.vmp_id for amp in amps}
+        vmps = VMP.objects.using(self.database_alias).filter(id__in=codes)
+
+        # exclude null VTM-VMP relationships (i.e. VMPs with no VTM)
+        return {(amp.vmp_id, amp.id) for amp in amps} | {
+            (vmp.vtm_id, vmp.id) for vmp in vmps if vmp.vtm_id
+        }
+
+    def descendant_relationships(self, codes):
+        vmps_from_vtms = VMP.objects.using(self.database_alias).filter(vtm__in=codes)
+
+        # get AMPs that have ancestor VMPs that are either in `codes`
+        # or are descendants of VTMs that are in `codes`
+        codes = set(codes) | {vmp.id for vmp in vmps_from_vtms}
+        amps_from_vmps = AMP.objects.using(self.database_alias).filter(vmp_id__in=codes)
+
+        # exclude null VTM-VMP relationships (i.e. VMPs with no VTM)
+        return (
+            {(vmp.vtm_id, vmp.id) for vmp in vmps_from_vtms}
+            | {(vmp.vtm_id, vmp.id) for vmp in vmps_from_vtms if vmp.vtm_id}
+            | {(amp.vmp_id, amp.id) for amp in amps_from_vmps}
+        )
```

Comment:
I think the first two parts of this set union return the same thing i.e. I don't think you need to exclude the null VTM-VMPs when doing descendants. By definition `vmps_from_vtms` is just vmps with a vtm_id, so the second statement is redundant. Another way of looking at it, the parent codes are either going to be VTMs or VMPs - if all VTMs have VMP children, and all VMPs have AMP children, then it's all fine.